### DISCUSSION TITLE
Sistema Angular e Spring para homologar

### DISCUSSION THREAD

**Discussion by guilhermecostalopes** [frontendbr/forum#1043] [Perguntas e Respostas]
Temos uma VM na azure.

Gerei o build de homologação e coloquei no servidor.

Gerou os arquivos e o index.html assim

    <!doctype html>
    <html lang="en">
    <head>
      <meta charset="utf-8">
      <title>DESIF</title>
      <base href="/">

      <link href="https://fonts.googleapis.com/icon?family=Material+Icons" rel="stylesheet">
      <meta name="viewport" content="width=device-width, initial-scale=1">
      <link rel="icon" type="image/x-icon" href="favicon.ico">
    </head>
    <body>

      <app-root>Carregando ...</app-root>

    <script type="text/javascript" src="inline.bundle.js"></script><script type="text/javascript" src="polyfills.bundle.js"></script><script type="text/javascript" src="styles.bundle.js"></script><script type="text/javascript" src="vendor.bundle.js"></script><script type="text/javascript" src="main.bundle.js"></script></body>
    </html>

No servidor Tomcat. na pasta **Program Files\Apache Software Foundation\Tomcat 8.5\webapps**, criei a pasta **dist**.

Ao entrar com o endereço: **http://endereco:8080/dist**, ele vai a página correta, ao carregar muda o endereço para **http://endereco:8080/#/login**, o que é correto, mas imaginei que seria: **http://endereco:8080/dist/#/login**

Se aperto **F5** ou **CTRL + F5**, ele vai para inicial do TOMCAT.

Tenho que configurar algo ?

No Angular, tenho um arquivo que tem o endereço do Servidor, no caso é **http://localhost:8080/sistema*, para não ter que ficar mudando toda vez que subir para homologação, tem como o angular fazer isto automático, por alguma variável ?

Primeira vez que faço um deploy.

Existe alguma maneira melhor ?

![servidor](https://user-images.githubusercontent.com/24395029/39022488-2a5684de-440c-11e8-8508-c5eac559be22.png)

**Comment by samuelsilvadev** [comment]
@guilhermecostalopes eae mano, tudo tranquilo?

Ta vendo um atributo `<base href="/">` faz um teste, coloca um .(ponto) antes da barra. ficando assim
`<base href="./">`

Qualquer coisa chama ae

**Comment by guilhermecostalopes** [comment]
@samuelsilvadev , funcionou.

Mas porque já não faz o build corretamente ?

Ou tenho que configurar algo ?

**Comment by samuelsilvadev** [comment]
Na verdade quando se faz o build vc pode especificar qual será a BASE_URL da sua app. Por default vem a /.

Para alterar esse comportamento padrão você tem duas opções:

1. Mudar manualmente na sua app e mandar para seu controle de versão.
2. Usar a flag `ng build --base-href BASE_URL` quando for fazer o build

Você pode obter mais detalhes em [Doc angular-cli](https://github.com/angular/angular-cli/wiki/build#base-tag-handling-in-indexhtml)

Particularmente eu utilizo a primeira opção, mas cada caso é um caso.

Espero ter ajudado, e qualquer coisa, só chamar :smile:

**Comment by guilhermecostalopes** [comment]
Acho que entendi o que você disse. Vou testar aqui.

**Comment by guilhermecostalopes** [comment]
Tem como fazer uma variável para servidor de homologação, produção e desenvolvimento, pois os mesmos mudam, no angular ?

**Comment by samuelsilvadev** [comment]
Você pode fazer uso dos NPM scripts.

La no package.json na propriedade "scripts", **adicione mais** alguns scripts

``` 

"scripts": {
    "builddev": "ng build --base-href BASE_URL_DEV",
    "buildprod": "ng build --prod --base-href BASE_URL_PROD",
    "buildtest": "ng build --base-href BASE_URL_TEST"",`
  }

```
Ai você executa: `npm run builddev`,  `npm run buildprod`, `npm run buildtest`

Vê se funfa :smile:

**Comment by heberfomin** [comment]
Amigos, posso entrar nessa conversa e pedir a ajuda de voces?
Seguinte: Eu tenho um projeto Angular 6 que faz requisições de autenticação de login, senha, guarda de rotas, etc numa API num servidor cloud. Depois de apanhar um bocado com x-origin o server foi resolvido. Antes, eu tinha que usar um proxy.conf.json para acessar a api. Depois de resolvido o problema do cors no server ele passou a funcionar direitinho.
Quando eu fiz o ng build e o projeto foi servido em IIS 10 acontece o seguinte: Em ambiente de desenvolvimento (localhost:4200) acessa normal. A API recebe na request o seguinte:
Request {#42
```
  #json: ParameterBag {#24
    #parameters: array:2 [
      "username" => "admin"
      "password" => "*****"
    ]
  }
```
```
Quando tento a mesma coisa em ambiente de produção a API recebe assim:
Request {#42
  #json: ParameterBag {#24
    #parameters: []
  }

```
Não mudei nada no codigo do angular. Somente fiz o ng build. 
Esse projeto está num IIS e eu uso o seguinte web.config nele:

```
<configuration>
<system.webServer>
  <rewrite>
    <rules>
      <rule name="Redirect all requests" stopProcessing="true">
        <match url=".*" />
        <conditions logicalGrouping="MatchAll">
            <add input="{REQUEST_FILENAME}" matchType="IsFile" negate="true" />
            <add input="{REQUEST_FILENAME}" matchType="IsDirectory" negate="true" />
        </conditions>
        <action type="Rewrite" url="/shippments/" />
        <!--<action type="Rewrite" url="/" />-->
        </rule>
    </rules>
  </rewrite>
</system.webServer>
</configuration>
```

Conseguiriam me ajudar?
Desculpe a intromissão.
Agradeço antecipadamente.
